resource_detectors/lib/opentelemetry/resource/detectors/azure.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'net/http'
+
+module OpenTelemetry
+  module Resource
+    module Detectors
+      # Azure contains detect class method for determining Azure environment resource attributes
+      module Azure
+        extend self
+
+        AZURE_METADATA_URI = 'http://169.254.169.254/metadata/instance/compute?api-version=2019-08-15'
+
+        def detect
+          metadata = azure_metadata
+          resource_attributes = {}
+
+          unless metadata.nil?
+            resource_attributes[OpenTelemetry::SemanticConventions::Resource::CLOUD_PROVIDER] = 'azure'
+            resource_attributes[OpenTelemetry::SemanticConventions::Resource::CLOUD_ACCOUNT_ID] = metadata['subscriptionId']
+            resource_attributes[OpenTelemetry::SemanticConventions::Resource::CLOUD_PLATFORM] = cloud_platform(metadata['provider'])
+            resource_attributes[OpenTelemetry::SemanticConventions::Resource::CLOUD_REGION] = metadata['location']
+            resource_attributes[OpenTelemetry::SemanticConventions::Resource::CLOUD_AVAILABILITY_ZONE] = metadata['zone']
+
+            resource_attributes[OpenTelemetry::SemanticConventions::Resource::HOST_ID] = metadata['vmId']
+            resource_attributes[OpenTelemetry::SemanticConventions::Resource::HOST_IMAGE_ID] = metadata.dig('storageProfile', 'imageReference', 'id')
+            resource_attributes[OpenTelemetry::SemanticConventions::Resource::HOST_TYPE] = metadata['vmSize']
+            resource_attributes[OpenTelemetry::SemanticConventions::Resource::HOST_NAME] = metadata['name']
+          end
+
+          resource_attributes.delete_if { |_key, value| value.nil? || value.empty? }
+          OpenTelemetry::SDK::Resources::Resource.create(resource_attributes)
+        end
+
+        private
+
+        def azure_metadata
+          response = Net::HTTP.get(URI(AZURE_METADATA_URI).to_s, { 'Metadata' => 'true' })
+          return unless response.code == '200'
+
+          JSON.parse(response.body)
+        rescue StandardError
+          nil
```

Comment:
Would it be sensible to log an error here to give folks a bread crumb into why the resource detector was unable create any attributes?

Reply:
Since we are talking about improving this packages experience: I think it might also be interesting to refactor a little bit to have more resource detectors:
* Azure 
* AWS
* GCP
* Generic (things like host_name, os_type)
* Container (container_name, container_image, container_id, container_tag)
* ...

I'm not sure yet how `Generic` will play when it should be overwritten by a more specific resource detector; and I do not know how that will work for faas workloads. I do think however that it would be wise to split that of in a separate PR(s).

Back on-topic:
From my point of view; I would not be opposed to log something if the resource detector could not create attributes. I think it can be easily overcome by just being specific in the configuration.
When using the `Autodetector.detect`; when it logs an error; the user can update the configuration accordingly.